### PR TITLE
KAFKA-9435: DescribeLogDirs automated protocol

### PR DESCRIPTION
**Pull request by tombentley**
Also add version 2 to make use of flexible versions, per KIP-482.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by tombentley** [comment]
@cmccabe are you able to review this?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 1]
```diff
@@ -2243,21 +2245,36 @@ public DescribeReplicaLogDirsResult describeReplicaLogDirs(Collection<TopicParti
...
-                partitionsByBroker.put(replica.brokerId(), new HashSet<>());
-            partitionsByBroker.get(replica.brokerId()).add(new TopicPartition(replica.topic(), replica.partition()));
+            DescribeLogDirsRequestData requestData = partitionsByBroker.get(replica.brokerId());
+            if (requestData == null) {
+                requestData = new DescribeLogDirsRequestData();
+                partitionsByBroker.put(replica.brokerId(), requestData);
+            }
+            DescribableLogDirTopic describableLogDirTopic = requestData.topics().find(replica.topic());
+            if (describableLogDirTopic == null) {
+                ArrayList<Integer> v = new ArrayList<>();
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: The left side should be `List`

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java [hunk 2]
```diff
@@ -49,105 +43,58 @@
...
-     */
-    private static final Schema DESCRIBE_LOG_DIRS_REQUEST_V1 = DESCRIBE_LOG_DIRS_REQUEST_V0;
-
-    public static Schema[] schemaVersions() {
-        return new Schema[]{DESCRIBE_LOG_DIRS_REQUEST_V0, DESCRIBE_LOG_DIRS_REQUEST_V1};
-    }
-
-    private final Set<TopicPartition> topicPartitions;
+    private final DescribeLogDirsRequestData data;
+    private final short version;
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java]: Like in https://github.com/apache/kafka/pull/7957 let's remove this field

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 1 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Also can we find a better name? What about `partitionIndex`?

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks for the PR. I've made a very quick first pass and left a few comments

**Comment by tombentley** [comment]
@mimaison done, if you want to do another pass.

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 3]
```diff
@@ -446,7 +446,11 @@ class RequestQuotaTest extends BaseRequestTest {
           new AlterReplicaLogDirsRequest.Builder(Collections.singletonMap(tp, logDir))
 
         case ApiKeys.DESCRIBE_LOG_DIRS =>
-          new DescribeLogDirsRequest.Builder(Collections.singleton(tp))
+          val data = new DescribeLogDirsRequestData()
+          data.topics().add(new DescribeLogDirsRequestData.DescribableLogDirTopic()
+            .setTopic(tp.topic())
+            .setPartitionIndex(Collections.singletonList(tp.partition())))
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala]: We can remove a bunch of brackets in this statement

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 4]
```diff
@@ -671,25 +671,36 @@ class ReplicaManager(val config: KafkaConfig,
...
 
       } catch {
         case _: KafkaStorageException =>
-          (absolutePath, new LogDirInfo(Errors.KAFKA_STORAGE_ERROR, Map.empty[TopicPartition, ReplicaInfo].asJava))
+          new DescribeLogDirsResponseData.DescribeLogDirsResult().setLogDir(absolutePath)
+            .setErrorCode(Errors.KAFKA_STORAGE_ERROR.code)
         case t: Throwable =>
           error(s"Error while describing replica in dir $absolutePath", t)
-          (absolutePath, new LogDirInfo(Errors.forException(t), Map.empty[TopicPartition, ReplicaInfo].asJava))
+          new DescribeLogDirsResponseData.DescribeLogDirsResult().setLogDir(absolutePath)
```
Comment (mimaison) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: I think it's slightly more readable if each setter is called on a new line. There're a few places where `setLogDir()` is called inline in this file.

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java [hunk 5]
```diff
@@ -18,172 +18,80 @@
...
 
-            logDirInfos.put(logDir, new LogDirInfo(error, replicaInfos));
-        }
+    public DescribeLogDirsResponse(Struct struct, short version) {
+        this.data = new DescribeLogDirsResponseData(struct, version);
     }
 
     /**
      * Constructor for version 0.
      */
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java]: Is this comment still acurate?

**Review by mimaison — CHANGES_REQUESTED** [review]
I've made a quick pass and left a few comments.
Can you also rebase on trunk?

**Comment by tombentley** [comment]
@mimaison fixed review comments and rebased.

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java [hunk 6]
```diff
@@ -49,105 +43,55 @@
...
-                for (Object partitionObj : topicStruct.getArray(PARTITIONS_KEY_NAME)) {
-                    int partition = (Integer) partitionObj;
-                    topicPartitions.add(new TopicPartition(topic, partition));
-                }
-            }
-        }
+        this.data = new DescribeLogDirsRequestData(struct, version);
     }
 
     // topicPartitions == null indicates requesting all partitions, and an empty list indicates requesting no partitions.
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java]: This comment can now be removed

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 7]
```diff
@@ -50,7 +51,7 @@ import org.apache.kafka.common.record._
 import org.apache.kafka.common.replica.PartitionView.DefaultPartitionView
 import org.apache.kafka.common.replica.ReplicaView.DefaultReplicaView
 import org.apache.kafka.common.replica.{ClientMetadata, _}
-import org.apache.kafka.common.requests.DescribeLogDirsResponse.{LogDirInfo, ReplicaInfo}
+import org.apache.kafka.common.message.DescribeLogDirsResponseData._
```
Comment (mimaison) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: Can we put this with the other import from the same package

**Review by mimaison — CHANGES_REQUESTED** [review]
I've made another pass and spotted a couple of tiny issues. Happy to merge once fixed

**Comment by tombentley** [comment]
@mimaison done

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java [hunk 8]
```diff
@@ -49,105 +43,54 @@
                     TOPIC_NAME,
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java]: This V0 schema should be deleted too

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java [hunk 9]
```diff
@@ -49,105 +43,54 @@
...
-                    topicPartitions.add(new TopicPartition(topic, partition));
-                }
-            }
-        }
+        this.data = new DescribeLogDirsRequestData(struct, version);
     }
 
     // topicPartitions == null indicates requesting all partitions, and an empty list indicates requesting no partitions.
-    public DescribeLogDirsRequest(Set<TopicPartition> topicPartitions, short version) {
+    public DescribeLogDirsRequest(DescribeLogDirsRequestData data, short version) {
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsRequest.java:55]: The comment above is out of date

**Review by mimaison — CHANGES_REQUESTED** [review]
I've left a couple of new comments for issues we should fix

**Comment by tombentley** [comment]
Sloppy! Thanks, now fixed.

**Review by mimaison — APPROVED** [review]

**Pull request merged**